WIP: switch to libevent for node socket handling

theuni opened the pull request:
Not yet ready for review. This can be considered a staging area. Chunks of this will be PR'd until only the actual libevent switch-over is ready, at which point this PR should be ready for review.

These changes remove our old select() loop for socket handling in favor of libevent, which uses epoll/kqueue/etc. as a back-end. In addition to being faster and more efficient, this allows us to drop some annoying restrictions, namely that select can only handle 1024 sockets in practice on most systems.

Note that this does _not_ yet make the switch to libevent for outgoing connections, that work is happening in parallel, and will be easier to merge after this.

Also, for any reviewers, several of these commits would individually introduce some regressions or slow-downs, but they've been split up in order to clarify why some of the changes are being made.

Depends on:
- [x] #10663 
- [x] #11363

Todo:
- Add a ton of documentation
- RAII the libevent structures
- Add some tests where possible

theuni committed:
net: when de-activating p2p, set nodes to be disconnected rather than closing the socket immediately

Messing with the socket from a non-network thread adds unnecessary complexity.

theuni committed:
net: Add a Disconnect function for nodes

For now, this just sets fDisconnect like before. But a function will be
necessary later.

theuni committed:
net: add an EnableReceive function so that the message handler can trigger it

for now this just disables fPauseRecv, but it will need to be a function later.

theuni committed:
net: get rid of optimistic send and the socket lock

The socket lock only existed because it was possible to trigger an immediate
disconnect via threads other than the socket handler.

The optimistic send was the last offender.

theuni committed:
net: move socket receiving to its own function

theuni committed:
net: narrow the scope for cs_vRecv to shared variables only

theuni committed:
net: fix possibly skipped decrefs at shutdown

theuni committed:
net: move inactivity checking to its own function

theuni committed:
net: pass the socket/whitelist args directly to AcceptConnection

Subsequent changes will turn AcceptConnection into a callback where the socket
is one of the supplied params.

theuni committed:
net: add an event handler for future read/write callbacks

For now this just handles read/write/error/timeout conditions for each node
serially, but OnEvents will soon be invoked as a callback when socket status
changes.

theuni committed:
net: hook up minimal libevent usage for connection acceptance

theuni committed:
net: switch to libevent for node socket handling

theuni committed:
net: Only close the socket in once place

By only allowing the socket to be closed and the handle to be invalidated in
one place, it becomes much easier to reason about the status at any given time.

theuni committed:
net: optimization: use scatter/gather to send up to 16 messages at once

theuni commented:
I believe that the proxy_test failure is a false-positive.

Looks like the issue is that disconnection happens so quickly now after an error that the testnode fails.

sipa commented on src/net.cpp:
```diff
@@ -2397,6 +2370,9 @@ bool CConnman::Start(CScheduler& scheduler, const Options& connOptions)
     }
+    m_event_timeout = timeval{60, 0};
+    const timeval* common_timeout = event_base_init_common_timeout(m_event_base, &m_event_timeout);
+    memcpy(&m_event_timeout, common_timeout, sizeof(m_event_timeout));
```
The documentation on `event_base_init_common_timeout` says "To do this, call this function with the common duration. It will return a pointer to a different, opaque timeout value. (Don't depend on its actual contents!)".

I'm not sure whether that means you need to use the exact pointer returned, or just a copy of its value (as this code does).

theuni replied:
Good eye! I wasn't sure how to store this properly either. This usage is taken from the official documentation. I'll add a comment about its sanity and a link to the docs.

sipa commented on src/net.cpp:
```diff
@@ -456,7 +471,14 @@ void CConnman::DumpBanlist()
 void CNode::CloseSocketDisconnect()
 {
-    Disconnect();
+    if (m_write_event) {
```
I would prefer it if you could have some RAII wrappers around the libevent structures before adding all this.

theuni replied:
Agreed, see PR description :)

theuni commented on src/net.cpp:
```diff
@@ -454,7 +454,7 @@ void CConnman::DumpBanlist()
 void CNode::CloseSocketDisconnect()
 {
-    fDisconnect = true;
+    Disconnect();
```
Probably so, will give it a shot.

theuni commented on src/net.cpp:
```diff
@@ -914,12 +914,7 @@ size_t CConnman::SocketSendData(CNode *pnode) const
         assert(data.size() > pnode->nSendOffset);
-        int nBytes = 0;
-        {
-            if (pnode->hSocket == INVALID_SOCKET)
```
I believe all of these are gone by the end of the change set, the removal may make more sense in an earlier commit, though.

theuni commented on src/net.cpp:
```diff
@@ -1205,11 +1302,14 @@ void CConnman::AcceptConnection(const SOCKET& listen_socket, bool whitelisted) {
-        CloseSocket(hSocket);
-        return;
+    const char* method = event_base_get_method(m_event_base);
+    if (!method || (strncmp(method, "select", 6) == 0)) {
```
It's only needed if select ends up being used, which should be very unlikely on most platforms. I'll add a comment here, and print something to the log so that it's clear which back-end is being used.

laanwj replied:
We could just bail out if select is being used. But not sure what platforms would be affected.

laanwj commented:
Concept ACK, woohoo!

jimpo commented on src/net.cpp:
```diff
@@ -841,36 +872,126 @@ const uint256& CNetMessage::GetMessageHash() const
+            }
 
+        }
+        WakeMessageHandler();
```
Should this be moved into the above `if` block? (It used to be inside the `if (notify)` statement)

theuni replied:
Yes, thank you.

sipa commented on src/net.cpp:
```diff
@@ -1150,7 +1150,7 @@ void CConnman::AcceptConnection(const ListenSocket& hListenSocket) {
     }
 
-    bool whitelisted = hListenSocket.whitelisted || IsWhitelistedRange(addr);
+    whitelisted = whitelisted || IsWhitelistedRange(addr);
```
> ||=

I don't think such an operator exists in C++.

jimpo replied:
Oops, you're right. Should have checked.

theuni commented on src/net.cpp:
```diff
@@ -1150,7 +1150,7 @@ void CConnman::AcceptConnection(const ListenSocket& hListenSocket) {
     }
 
-    bool whitelisted = hListenSocket.whitelisted || IsWhitelistedRange(addr);
+    whitelisted = whitelisted || IsWhitelistedRange(addr);
```
I can say for certain that the operator doesn't exist because I tried it first :)

laanwj replied:
Hehe I also remember it doesn't exist - I looked it up a long time ago and IIRC the rationale is that's because `|=` is the same when using only booleans. In any case this is ok as it is.

theuni committed:
scripted-diff: Call Disconnect rather than setting fDisconnect directly

-BEGIN VERIFY SCRIPT-
sed -i 's/fDisconnect = true/Disconnect()/' src/net.cpp src/net_processing.cpp
-END VERIFY SCRIPT-

theuni committed:
net: move connection parameters into a struct and pass them all at once

theuni committed:
net: merge duplicate checks

theuni committed:
net: create callbacks for new connections and outgoing failures

theuni committed:
net: combine OpenNetworkConnection and ConnectNode

theuni committed:
net: use OnNewConnection for incoming connections as well

theuni committed:
net: add RAII helper for outgoing connection failure callback

theuni committed:
Net: handle all new connections on the socket handler thread

This ensures that outgoing connections don't race with event teardown at
shutdown.

theuni committed:
net: split socket creation out of connection

promag commented:
Needs rebase.

fanquake commented:
Removing "Up for grabs", given we are now in the process of trying to remove libevent.
